### PR TITLE
L059 quoted identifiers bug

### LINKED ISSUE #2611: Potential Bug with TSQL Dialect QUOTED_IDENTIFIER keyword
## Expected Behaviour

Setting the value for `QUOTED_IDENTIFIER` should not trigger the unnecessary quoted_identifier rule L059 for the following tsql code.
```
SET ANSI_NULLS ON
GO

SET QUOTED_IDENTIFIER ON
GO






CREATE PROCEDURE [dbo].[usp_job_email_route_error]
	@in_debug INT = 0

AS

/****************************************************************************************************************************************************************************************

	Will send an email containing route error information

*****************************************************************************************************************************************************************************************
Changelog (User / Date / Comments):

****************************************************************************************************************************************************************************************/
```

## Observed Behaviour
Running `sqlfluff lint` on the attached `sample.txt` file will result in this output. 
```
$ sqlfluff lint ~/Desktop/sample.sql --config ~/source/repos/company/solution/database/.sqlfluff 
== [C:\Users\user\Desktop\sample.sql] FAIL
L:   4 | P:   5 | L059 | Unnecessary quoted identifier QUOTED_IDENTIFIER.
L:   5 | P:   3 |  PRS | Line 5, Position 3: Found unparsable section:
                       | '\n\n\n\n\n\n\nCREATE PROCEDURE [dbo].[usp_job_e...'
L:  24 | P:   1 | L009 | Files must end with a single trailing newline.
All Finished 📜 🎉!
```

Running `sqlfluff fix` will result in 'trimming' the keyword `QUOTED_IDENTIFIER`. 

I was getting a result that looks as such for line 4: `SET UOTED_IDENTIFIE ON`

## Steps to Reproduce
Save `sample.txt` to `.sql` file. 
Run `sqlfluff lint <file_name>`
Check line 4. Line 5 also showing parse error, but I am not convinced this is a bug.  

## Dialect
TSQL 

## Version
sqlfluff, version 0.10.0
Python 3.9.10

## Configuration
```
[sqlfluff]
dialect = tsql

[sqlfluff:indentation]
indented_joins = false
indented_using_on = true

[sqlfluff:rules]
indent_unit = tab
tab_space_size = 4
operator_new_lines = after
capitalisation_policy = upper
aliasing = explicit
allow_scalar = True
extended_capitalisation_policy = consistent
ignore_comment_clauses = True
ignore_comment_lines = True
max_line_length = 120
comma_style = leading
single_table_references = consistent
select_clause_trailing_comma = forbid
prefer_count_1 = True
multiline_newline = True
require_final_semicolon = True
allow_space_in_identifier = True
quoted_identifiers_policy = all
unquoted_identifiers_policy = all
exclude_rules = L025, L031, L061
```
[parse.txt](https://github.com/sqlfluff/sqlfluff/files/8048142/parse.txt)
[sample.txt](https://github.com/sqlfluff/sqlfluff/files/8048146/sample.txt)

### PR DESCRIPTION
**Pull request by tunetheweb**
### Brief summary of the change made

Fixes #2611

### Are there any other side effects of this change that we should be aware of?
Nope

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/rules/L059.py [hunk 1]
```diff
@@ -84,7 +84,13 @@ def _eval(self, context: RuleContext) -> Optional[LintResult]:
             identifier_contents = context.segment.raw[1:-1]
 
         # Ignore the segments that are not of the same type as the defined policy above.
-        if context.segment.name not in context_policy:
+        # Also TSQL has a keyword called QUOTED_IDENTIFIER which maps to the name so
+        # need to explicity check for that.
+        if (
+            context.segment.name not in context_policy
+            or context.segment.raw.lower() == "quoted_identifier"
+            or context.segment.raw.lower() == "naked_identifier"
```
Comment (barrywhart) [inline src/sqlfluff/rules/L059.py:92]:
```suggestion
            or context.segment.raw.lower() in ("quoted_identifier", "naked_identifier")
```

Comment (barrywhart) [reply src/sqlfluff/rules/L059.py:92]: I'm wondering about this code -- it's checking `raw`, but raw is the actual, rendered code. I'm surprised this isn't `name` or `type`.

File: src/sqlfluff/rules/L059.py [hunk 2]
```diff
@@ -106,6 +112,10 @@ def _eval(self, context: RuleContext) -> Optional[LintResult]:
         # Retrieve NakedIdentifierSegment RegexParser for the dialect.
         naked_identifier_parser = context.dialect._library["NakedIdentifierSegment"]
 
+        print("BARRY1:%s:" % (quoted_identifier_contents))
+        print("BARRY2:%s:" % (naked_identifier_parser.template))
+        print("BARRY3:%s:" % (naked_identifier_parser.anti_template))
+
```
Comment (barrywhart) [inline src/sqlfluff/rules/L059.py:118]:
```suggestion
```

**Review by barrywhart — CHANGES_REQUESTED** [review]

File: src/sqlfluff/rules/L059.py (hunk 1 above)
Comment (tunetheweb) [reply src/sqlfluff/rules/L059.py]: The problem is for a keyword the name is the same as the raw. Hence this issue.

We could check the name and if it’s a keyword, but thought above check was more comprehensive.

Comment (tunetheweb) [reply src/sqlfluff/rules/L059.py]: BTW the full `if` statement is:

```py
        if (
            context.segment.name not in context_policy
            or context.segment.raw.lower() in ("quoted_identifier", "naked_identifier")
        ):
```

That is, if the name is not one we're interested in OR the raw value is literally the value `QUOTED_IDENTIFIER` (which is a `name` in SQLFluff parser language BUT ALSO an actual KEYWORD in T-SQL 😱), then it's not one we care about.

**Review by barrywhart — APPROVED** [review]

**Pull request merged**